### DISCUSSION TITLE
Improve Composer Dump autoload performance

### DISCUSSION THREAD

**Discussion by jordikroon** [craftcms/cms#12583] [General]
Hi guys,

I have hesitated a long time before posting this here because I don't really know what the cause is of this issue and/or if it is Yii2 related, or Craft.

Have been using Craft for a while, and the single thing that I find really annoying is the time it takes for composer to (re)generate autoloading files after a post install or update.

It seems significant slower compared to other frameworks like Symfony or (obviously) using bare composer and installing a few packages only.

If this could be improved in some way, that would help our development processes a lot. It would also save us a lot of time during our monthly maintenance round. 

Thank you,
Jordi

**Comment by bencroker** [comment]
Are you sure you're running Composer 2? Check the version with `composer -V`. Also, are you running Composer locally on your machine or in Docker or another development environment? And how slow are talking, exactly?

**Reply by jordikroon** [reply]
Hey Ben,

We are running Composer 2.5.1 (2022-12-22 15:33:54 build) within a docker environment. Running on MacOS.

And to put things into perspective. We ran `composer dump` on both a Symfony project, and a Craft project. Under the hood the same Dockerfile configuration is used.

**Craft 4:**
```
root@a18162ecd294:/app# time composer dump
Generating optimized autoload files
Generated optimized autoload files containing 7597 classes

real	0m56.342s
user	0m50.684s
sys	0m1.417s
```

**Symfony:**
```
root@756da4bcf292:/app# time composer dump
Generating autoload files
Generated autoload files

real	0m2.669s
user	0m2.036s
sys	0m0.318s
```

Now the thing I am noticing is that Craft generated an optimised autoload file, when I run `composer dump -o` on both projects I get different results but the difference is still very much noticeable.

**Craft 4:**
```
root@a18162ecd294:/app# time composer dump -o
Generating optimized autoload files
Generated optimized autoload files containing 7597 classes

real	0m51.733s
user	0m47.173s
sys	0m1.319s
```

**Symfony:**
```
root@756da4bcf292:/app# time composer dump -o
Generating optimized autoload files
Generated optimized autoload files containing 8994 classes

real	0m18.319s
user	0m14.779s
sys	0m1.044s
```

It is not the end of the world, but if an improvement could be made. That would be nice :)

**Reply by bencroker** [reply]
I'm seeing sub 10s results in a DDEV setup.

```
time ddev composer dump
Generating optimized autoload files
Generated optimized autoload files containing 10170 classes
ddev composer dump  0.07s user 0.09s system 1% cpu 9.521 total
```

Due to the way PSR-4 and PSR-0 autoloading rules are set up, Composer needs to check the filesystem before resolving a classname conclusively. Since Docker can be slow at accessing the filesystem, it likely depends on your Docker config and machine specs. I'm currently running Docker 4.16.2 with VirtioFS enabled (which improves I/O performance) on MacOS with an M1 Pro chip.

**Reply by jordikroon** [reply]
@bencroker I have been analysing a bit further. My conclusion is that it has nothing to do with the Craft composer plugin nor Yii2 composer (disabled them both).

However I discover that the slowness is caused by composer generating "optimised" autoloading files. We most certainly have this config option set because it is defined in the [Craft Boilerplate (see)](https://github.com/craftcms/craft/blob/main/composer.json#L40).

We are considering removing this property since this is already defined in our deployment process. 

I am curious to know what the drawback related to Craft CMS is besides a minor performance drawback. Composer strongly discourages this setting for development environments as written here: https://getcomposer.org/doc/articles/autoloader-optimization.md

 > Note: You should not enable any of these optimizations in development as they all will cause various problems when  
 > adding/removing classes. The performance gains are not worth the trouble in a development setting.

On a fresh installation we see 10s results with the optimize flag enabled, and 0.8s with optimize disabled. On one of our largest installations it various between 20s and a full minute and also around 0.8-1.2s with optimise disabled.

For that reason our issue has been resolved, but it might be good to remove the ` "optimize-autoloader": true` from the default installation.

**Comment by bencroker** [comment]
The default Craft `composer.json` enables class map generation using `"optimize-autoloader": true` in preparation for production environments. It may be worth asking P&T to change this, but my guess is that making Composer autoloader performant takes higher priority than making it developer friendly, out-of-the-box at least. And of course you can always remove that config option and just ensure you run `composer dump -o` in your production environment.

**Reply by brandonkelly** [reply]
Yeah in general your classmap should be optimized. And as long as you don’t run with `-a` (authoritative), Composer can still autoload classes that aren’t included in the classmap.
